### PR TITLE
[ISSUE #8275] Fix ClassCastException: io.grpc.ClientInterceptors$InterceptorChannel cannot be cast to io.grpc.ManagedChannel

### PR DESCRIPTION
**Pull request by ZhengBing520**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

https://github.com/alibaba/nacos/issues/8275

## Brief changelog

Get the created channel directly

## Verifying this change

no output error log

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Review by ZhengBing520 — COMMENTED** [review]
Adjust the channel acquisition method

**Comment by ZhengBing520** [comment]
image

**Comment by KomachiSion** [comment]
1. your commit user is not your github user, you need change it.
2. Waiting CLA page load finished, and roll down the end and sign agree.

image
image

**Comment by ZhengBing520** [comment]
> 1. your commit user is not your github user, you need change it.
> 2. Waiting CLA page load finished, and roll down the end and sign agree.
> 
> image image

Thank you, it's taken care of

**Comment by KomachiSion** [comment]
@shiyiyue1102 Please help review this PR.

**Comment by KomachiSion** [comment]
@shiyiyue1102 Please help review this PR.

File: common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java [hunk 1]
```diff
@@ -151,6 +151,30 @@ private RequestGrpc.RequestFutureStub createNewChannelStub(String serverIp, int
...
+
+    /**
+     * create a new channel with specific server address.
+     *
+     * @param serverIp   serverIp.
+     * @param serverPort serverPort.
+     * @return if server check success,return a non-null channel.
+     */
+    private ManagedChannel createNewManagedChannel(String serverIp, int serverPort) {
+        ManagedChannelBuilder<?> o = ManagedChannelBuilder.forAddress(serverIp, serverPort).executor(grpcExecutor)
```
Comment (Oliverwqcwrw) [inline common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java:173]:
```suggestion
        ManagedChannelBuilder<?> managedChannelBuilder = ManagedChannelBuilder.forAddress(serverIp, serverPort).executor(grpcExecutor)
```

File: common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java [hunk 2]
```diff
@@ -151,6 +151,30 @@ private RequestGrpc.RequestFutureStub createNewChannelStub(String serverIp, int
...
+                .compressorRegistry(CompressorRegistry.getDefaultInstance()).decompressorRegistry(DecompressorRegistry.getDefaultInstance())
+                .maxInboundMessageSize(getInboundMessageSize()).keepAliveTime(keepAliveTimeMillis(), TimeUnit.MILLISECONDS).usePlaintext();
+        return o.build();
```
Comment (Oliverwqcwrw) [inline common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java:176]:
```suggestion
        return managedChannelBuilder.build();
```

**Review by Oliverwqcwrw — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java (hunk 1 above)
Comment (ZhengBing520) [reply common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java:173]: hi，Whether line 141 needs to be deleted

Comment (Oliverwqcwrw) [reply common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java:173]: I think line 142 needs to be deleted

Comment (ZhengBing520) [reply common/src/main/java/com/alibaba/nacos/common/remote/client/grpc/GrpcClient.java:173]: Ok, submit it now

**Comment by ZhengBing520** [comment]
hello @Oliverwqcwrw  Submitted, please check

**Review by shiyiyue1102 — APPROVED** [review]

**Review by KomachiSion — CHANGES_REQUESTED** [review]
Can you add some unit test for your changes？

**Comment by ZhengBing520** [comment]
> Can you add some unit test for your changes？

After adjusting the code, local startup is normal.  How to add unit test?    Can you give me an example
thanks for you

**Comment by KomachiSion** [comment]
There are many unit test in project, you can see any `src/test` in modules.

**Comment by ZhengBing520** [comment]
> There are many unit test in project, you can see any `src/test` in modules.

Do I only need to provide the two methods I add？

**Comment by KomachiSion** [comment]
> > There are many unit test in project, you can see any `src/test` in modules.
> 
> Do I only need to provide the two methods I add？

We hope you can add the cases to cover your changes, which can test the class has same behavior after your changes.

**Comment by KomachiSion** [comment]
CI can't pass after your changes.

**Comment by ZhengBing520** [comment]
> CI can't pass after your changes.

Let me check that I removed a referenceless method on my second commit ([GrpcClient.java] line 141)

**Comment by ZhengBing520** [comment]
> CI can't pass after your changes.

I have successfully executed the second time locally, shall we execute it again  ->  Continuous Integration / unix (ubuntu-latest, 8) (pull_request) 
image

**Comment by Oliverwqcwrw** [comment]
https://github.com/alibaba/nacos/runs/6416290642?check_suite_focus=true
There may have some problem,You should check it and solve it

**Comment by ZhengBing520** [comment]
> https://github.com/alibaba/nacos/runs/6416290642?check_suite_focus=true There may have some problem,You should check it and solve it

I try to pull the latest code

**Comment by ZhengBing520** [comment]
> > https://github.com/alibaba/nacos/runs/6416290642?check_suite_focus=true There may have some problem,You should check it and solve it
> 
> I try to pull the latest code

hello @KomachiSion @Oliverwqcwrw Is it all right now, do you need me to rebase it？

**Comment by ZhengBing520** [comment]
image

**Comment by ZhengBing520** [comment]
hello @KomachiSion I resubmitted a PR：https://github.com/alibaba/nacos/pull/8374

**Comment by KomachiSion** [comment]
> hello @KomachiSion I resubmitted a PR：#8374

OK, I will close this PR and move to new PR discuss and review.

**Comment by ZhengBing520** [comment]
Thanks for the guidance. I've taken care of step two




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "alibaba/nacos"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年4月29日(星期五) 中午11:28
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [alibaba/nacos] [ISSUE #8275] Fix ClassCastException: io.grpc.ClientInterceptors$InterceptorChannel cannot be cast to io.grpc.ManagedChannel (PR #8276)





  
your commit user is not your github user, you need change it.
 
Waiting CLA page load finished, and roll down the end and sign agree.

**Pull request closed without merging**